### PR TITLE
merge #16117, #18358, #17383, #21052, #14424, #15159, #14689, #14978, partial #16908, #14978, #13932: Auxillary Backports

### PR DESCRIPTION
**Pull request by kittywhiskers**
### Notes

~~bitcoin#20788 was originally having test failures and the lack of bitcoin#14624 was the cause. Tests run fine now.~~ Scrapped from this PR

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
Windows build borked

**Comment by kittywhiskers** [comment]
Rebased and updated

**Comment by PastaPastaPasta** [comment]
Not compiling

File: src/net.cpp [hunk 1]
```diff
@@ -472,50 +473,52 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
...
+        if (!sock) {
             return nullptr;
         }
         std::string host;
         int port = default_port;
         SplitHostPort(std::string(pszDest), port, host);
-        connected = ConnectThroughProxy(proxy, host, port, hSocket, nConnectTimeout, nullptr);
+        bool proxyConnectionFailed;
+        connected = ConnectThroughProxy(proxy, host, port, *sock, nConnectTimeout,
+                                        proxyConnectionFailed);
```
Comment (UdjinM6) [inline src/net.cpp:511]: `proxyConnectionFailed` related changes are from 17754

File: src/netbase.cpp [hunk 2]
```diff
@@ -245,21 +241,22 @@ enum class IntrRecvError {
  * or return False on error or timeout.
  * This function can be interrupted by calling InterruptSocks5()
  *
- * @param data Buffer to receive into
- * @param len  Length of data to receive
- * @param timeout  Timeout in milliseconds for receive operation
+ * @param data The buffer where the read bytes should be stored.
+ * @param len The number of bytes to read into the specified buffer.
+ * @param timeout The total timeout in milliseconds for this read.
```
Comment (UdjinM6) [inline src/netbase.cpp:246]: this is from some another PR

File: src/netbase.cpp [hunk 3]
```diff
@@ -268,23 +265,10 @@ static IntrRecvError InterruptibleRecv(uint8_t* data, size_t len, int timeout, c
         } else { // Other error or blocking
             int nErr = WSAGetLastError();
             if (nErr == WSAEINPROGRESS || nErr == WSAEWOULDBLOCK || nErr == WSAEINVAL) {
-                if (!IsSelectableSocket(hSocket)) {
-                    return IntrRecvError::NetworkError;
-                }
+                // Only wait at most maxWait milliseconds at a time, unless
+                // we're approaching the end of the specified total timeout
```
Comment (UdjinM6) [inline src/netbase.cpp:269]: same

File: src/netbase.cpp [hunk 4]
```diff
@@ -330,8 +314,25 @@ static std::string Socks5ErrorString(uint8_t err)
...
+ * @param sock The SOCKS5 proxy socket.
+ *
+ * @returns Whether or not the operation succeeded.
+ *
+ * @note The specified SOCKS5 proxy socket must already be connected to the
+ *       SOCKS5 proxy.
+ *
+ * @see <a href="https://www.ietf.org/rfc/rfc1928.txt">RFC1928: SOCKS Protocol
+ *      Version 5</a>
+ */
```
Comment (UdjinM6) [inline src/netbase.cpp:334]: same

File: src/netbase.cpp [hunk 5]
```diff
@@ -609,24 +614,38 @@ bool IsProxy(const CNetAddr &addr) {
...
+ * @param strDest The destination service to which to connect.
+ * @param port The destination port.
+ * @param sock The socket on which to connect to the SOCKS5 proxy.
+ * @param nTimeout Wait this many milliseconds for the connection to the SOCKS5
+ *                 proxy to be established.
+ * @param[out] outProxyConnectionFailed Whether or not the connection to the
+ *                                      SOCKS5 proxy failed.
+ *
+ * @returns Whether or not the operation succeeded.
+ */
```
Comment (UdjinM6) [inline src/netbase.cpp:631]: same

File: src/netbase.cpp [hunk 6]
```diff
@@ -609,24 +614,38 @@ bool IsProxy(const CNetAddr &addr) {
...
+bool ConnectThroughProxy(const proxyType& proxy, const std::string& strDest, int port, const Sock& sock, int nTimeout, bool& outProxyConnectionFailed)
```
Comment (UdjinM6) [inline src/netbase.cpp:632]: `*outProxyConnectionFailed` -> `&outProxyConnectionFailed` is from 17754

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
16908: missing changes in `test/functional/mempool_persist.py`
18358: see ebc239029c6e464c38cc1010755cfc387cef9a41
20788: lots of changes from other PRs

File: src/util/time.h [hunk 7]
```diff
@@ -12,6 +12,19 @@
...
+ * All durations should be using std::chrono and calling this should generally
+ * be avoided in code. Though, it is still preferred to an inline t.count() to
+ * protect against a reliance on the exact type of t.
+ *
+ * This helper is used to convert durations before passing them over an
+ * interface that doesn't support std::chrono (e.g. RPC, debug log, or the GUI)
+ */
+inline int64_t count_seconds(std::chrono::seconds t) { return t.count(); }
+inline int64_t count_microseconds(std::chrono::microseconds t) { return t.count(); }
+
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/util/time.h]: Merged in #18638

**Comment by PastaPastaPasta** [comment]
functional tests borked

**Comment by kittywhiskers** [comment]
Dunno what's happening with functional tests so #16908 has been slimmed to the bare essentials needed for successive commits

**Comment by PastaPastaPasta** [comment]
Needs rebase

**Comment by PastaPastaPasta** [comment]
Tests are failing

**Comment by PastaPastaPasta** [comment]
Marking as WIP due to the fact this PR has stalled. Please rebase, and mark as ready for review when this PR is RfR

**Comment by kittywhiskers** [comment]
Ready for review. Unrelated LLMQ tests failed

**Comment by PastaPastaPasta** [comment]
Test failures in feature_llmq_data_recovery.py are persistent. Does not seem unrelated.

**Comment by kittywhiskers** [comment]
> Test failures in feature_llmq_data_recovery.py are persistent. Does not seem unrelated.

Fixed, problem was isolated to 26599604e3739a37fe8d4fea932dd4074b7501eb

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
Pls see https://github.com/UdjinM6/dash/commits/pr4229

**Review by UdjinM6 — APPROVED** [review]
utACK

**Pull request merged**